R/render-table.R

```diff
@@ -37,23 +104,96 @@ renderTable <- function(expr, ..., env=parent.frame(), quoted=FALSE, func=NULL)
+    ## sure to left-align them (xtable right-aligns them by default, which
+    ## looks weird when the row.names are strings).
+    if (is.null(align)) {
+      cols <- if (isNumber(row.names(data))) "r" else "l"
+      for (i in seq_len(ncol(data))) {
+        cls <- class(data[,i])
+        if (cls == "numeric" || cls == "integer") cols <- paste0(cols, "r")
+        else cols <- paste0(cols, "l")
+      }
+    } else {
+      ## Case 2: if align!=NULL, check if it is only one character or a vector
+      ## and process it accordingly.
+      num_cols <- if (rownames) nchar(align) else nchar(align)+1
+      valid <- !grepl("[^lcr]", align)
+      if (num_cols == ncol(data)+1 && valid) {
+        cols <- if (rownames) align else paste0("r", align)
+      } else if (nchar(align) == 1 && valid) {
+        cols <- paste0(rep(align, ncol(data)+1), collapse="")
+      } else {
+        stop("`align` must contain only the characters `l`, `c` and/or `r` and have
+             length either equal to 1 or to the total number of columns")
+      }
+    }
+
+    # Call xtable with its (updated) args
+    xtable_args <- c(xtable_args, align=cols, digits=digits)
     xtable_res <- do.call(xtable, c(list(data), xtable_args))
 
     # Set up print args
     print_args <- list(
       xtable_res,
-      type = 'html',
-      html.table.attributes = paste('class="', htmlEscape(classNames, TRUE),
-                                    '"', sep='')
-    )
+      type='html',
+      include.rownames=rownames,
+      include.colnames=colnames,
+      NA.string=na,
+      html.table.attributes=paste0("class='", htmlEscape(classNames, TRUE), "' ",
+                                   "style='width:", validateCssUnit(width),";'"))
+
     print_args <- c(print_args, non_xtable_args)
 
-    return(paste(
+    # Capture the raw html table returned by print.xtable(), and store it in
+    # a variable for further processing
+    tab <- paste(
       utils::capture.output(
         do.call(print, print_args)
       ),
       collapse="\n"
-    ))
+    )
+
+    # All further processing concerns the table headers, so we don't need to run
+    # any of this if colnames=FALSE
+    if (colnames) {
+      # Make sure that the final html table has a proper header (not included
+      # in the print.xtable() default)
+      tab <- sub("<tr>", "<thead> <tr>", tab)
+      tab <- sub("</tr>", "</tr> </thead> <tbody>", tab)
+      tab <- sub("</table>", "</tbody> </table>", tab)
+
+      # Update the `cols` string (which stores the alignment of each column) so
+      # that it only includes the alignment for the table variables (and not
+      # for the row.names)
+      cols <- if (rownames) cols else substr(cols,2,nchar(cols))
+
+      # Create a vector whose i-th entry corresponds to the i-th table variable
+      # alignment (substituting "l" by "left", "c" by "center" and "r" by "right")
+      header_alignments <- c()
```

Comment:
I agree that it may be better to have `header_alignments <- character(0)`, although `header_alignments <- c()` (or, as you pointed out, its equivalent `header_alignments <- NULL`) also does the trick. I think it's nice to have `header_alignments <- c()` instead of the equivalent `header_alignments <- NULL` because it primes the reader for the fact that `header_alignments` will be a vector, but I agree that it is obscuring reality a little bit. Anyway, I changed it to `character(0)` for now.

I don't understand your second point though. How is this line not necessary? If the variable is not defined at all, I can't use `header_alignments[i]` in line 182, right? For example, here's a small console example:

```
> vec                 ## making sure `vec` is not defined
Error: object 'vec' not found
> vec[1] <- "hello"   ## trying to assign a value to the first slot of `vec`
Error in vec[1] <- "hello" : object 'vec' not found
```